src/CityOnPlanet.cpp

```diff
@@ -202,129 +305,354 @@ CityOnPlanet::~CityOnPlanet()
+
 		}
 	}
 
-	Aabb buildAABB;
-	for (std::vector<BuildingDef>::const_iterator iter = m_buildings.begin(), itEND = m_buildings.end(); iter != itEND; ++iter) {
-		buildAABB.Update((*iter).pos - p);
+	_genTimer.Stop();
+
+	Log::Verbose("\tCityOnPlanet: generated {} buildings in {}ms ( {} skipped, {} discarded, {} occupied, {} underwater, {} avg rolls / building )",
+		m_buildings.size(), _genTimer.milliseconds(), skippedCells, discardedCells, occupiedCells, underwaterCells, rarityRolls / double(m_buildings.size()));
+
+	_genTimer.SoftReset();
+
+	// Compute the total AABB of this city
+	Aabb totalAABB;
+	const vector3d &stationOrigin = station->GetPosition();
+
+	for (const auto &buildingInst : m_buildings) {
+		totalAABB.Update(buildingInst.pos - stationOrigin);
 	}
-	m_realCentre = buildAABB.min + ((buildAABB.max - buildAABB.min) * 0.5);
-	m_clipRadius = buildAABB.GetRadius();
-	AddStaticGeomsToCollisionSpace();
+
+	m_realCentre = totalAABB.min + ((totalAABB.max - totalAABB.min) * 0.5);
+	m_clipRadius = totalAABB.GetRadius();
+
+	_genTimer.Stop();
+
+	Log::Verbose("\tCityOnPlanet: generated AABB for city in {}ms", _genTimer.milliseconds());
+
+	// Release the memory once we're done generating and reset
+	m_gridBitset.reset();
+	m_gridPitch = 0;
+	m_gridLen = 0;
+}
+
+// Calculate the radius for this city based on the SystemBody's parameters
+void CityOnPlanet::CalcCityRadius(const SystemBody *body)
+{
+	double population = body->GetPopulation();
+	double atmo = body->GetAtmosOxidizing();
+
+	for (const CityRadiusDef &radii : reverse_container(m_cityType->sizeDefs)) {
+		bool isLast = (&radii == &m_cityType->sizeDefs.front());
+
+		if (population > radii.population && !isLast)
+			continue;
+
+		m_cityRadius = radii.baseSize + (atmo * radii.atmoSize) + (m_rand.Double() * radii.randomSize);
+		m_cityDensity = radii.density * 0.8;
+		break;
+	}
+}
+
+#ifdef __BIG_ENDIAN__
+#error "CityOnPlanet does not (yet) support big-endian architectures"
+#endif
+
+// Supports up to 32-cell wide buildings to ensure 8-byte aligned addressing in a single operation
+static inline uint64_t CalcBitmaskForGrid(uint32_t x, uint32_t xsize)
+{
+	uint32_t maskOffset = x & CityOnPlanet::CELLMASK;
+	uint64_t bitmask = uint64_t((1 << xsize) - 1) << maskOffset;
+	// note: need crossplatform bswap_64 here for big-endian compat
+	// we should be able to swap the bitmask rather than the in-memory bytes and have the same effect
+
+	return bitmask;
+}
+
+void CityOnPlanet::SetGridOccupancy(uint32_t x, uint32_t y, const uint8_t size[2])
+{
+	if (x + size[0] > m_citySize || y + size[1] > m_citySize)
+		return; // footprint would be off-grid, prevent writing outside the bitset
+
+	// City grid is stored as lsb-first bitset with one bit per grid cell;
+	// bit 1 of byte 1 is grid 0, while bit 8 of byte 4 is grid 31
+	// Support maximum size of 32 cells for a single building to enable fast 64-bit aligned operations
+
+	uint8_t *dataPtr = m_gridBitset.get() + (x & ~CELLMASK) / 8;
```

Comment:
Oh, by the way, why not just `>> 3` ?

Reply:
Compiler should fold it to `>> 3` anyways, and this is cleaner and easier to understand for future contributors who aren't versed in the dark art of C bitops.